### ISSUE TITLE
NPE in `observeNetworkConnectivity`

### ISSUE THREAD

**Issue by AllanWang** [pwittchen/ReactiveNetwork#209]
I'm getting the error:

`Attempt to invoke interface method 'java.util.Iterator java.util.List.iterator()' on a null object reference`

To make things simple, [here](http://crashes.to/s/c41455d77fc) is the crashlytics log

and [here](https://github.com/AllanWang/Frost-for-Facebook/blob/master/app/src/main/kotlin/com/pitchedapps/frost/activities/BaseActivity.kt) is my base activity.

All that is happening is that I ensure I dispose everything and then subscribe again on resume.

The error seems to be traced down to [`getNetworkInfo`](https://github.com/pwittchen/ReactiveNetwork/blob/RxJava2.x/library/src/main/java/com/github/pwittchen/reactivenetwork/library/rx2/Connectivity.java#L83). Perhaps there needs to be a check to ensure there are active networks before retrieving them?

Using 0.11.0

Edit: [Relevant](https://stackoverflow.com/questions/16397494/connectivitymanager-getactivenetworkinfo-null-pointer-exception). Seems like it may be an internal issue, given that you have null checks after the networks are retrieved.

**Comment by pwittchen** [comment]
There are null checks for `NetworkInfo`. It looks like `Context` or `ConnectivityManager` is null for some reason. I don't know if it possible for the `ConnectivityManager`, but we can add a null check for this object. `Context` may be null only in the case when you pass null as a `Context` to the library.

**Comment by AllanWang** [comment]
I'm passing `applicationContext` like the sample so that should be fine. It does look like `ConnectivityManager` is null in this case. I'm not sure if this is a physical device or an emulated one given that there's no information on RAM, but perhaps it will be worthwhile to add a null check there?

**Comment by pwittchen** [comment]
I'm not sure if it may happen that `ConnectivityManager` becomes `null`, but according to this discussion that's the only possible source of this NPE here. Due to this fact, we can add a null check for it.

**Comment by AllanWang** [comment]
For `getSystemService`:

`@return The service or null if the name does not exist.`

So I guess it is possible. Thanks

**Comment by pwittchen** [comment]
Fixed. Bug fix will be available in the next release.

**Comment by consp1racy** [comment]
@AllanWang The stacktrace is great help. 

@AllanWang @pwittchen Guys, I just dug in AOSP a bit and there's no way ConnectionManager is null. See `android.app.ContextImpl:289` in SDK 15 and `android.app.SystemServiceRegistry:790` in SDK 25. Whatever happens, there always is a non-null object.

If you take a look at the exception

    ...
    android.net.IConnectivityManager$Stub$Proxy.getActiveNetworkInfo (IConnectivityManager.java:1058)
    android.net.ConnectivityManager.getActiveNetworkInfo (ConnectivityManager.java:658)
    com.github.pwittchen.reactivenetwork.library.rx2.Connectivity.getNetworkInfo (Connectivity.java:86)
    ...

you can see that ConnectivityManager indeed was not null.

The error happens somewhere on the way from `android.net.ConnectivityManager` to `com.android.server.ConnectivityService`. Unfortunately IPC in AOSP is above my paygrade.

The only way to suppress this is to wrap `ConnectivityManager.getActiveNetworkInfo` in try-catch looking for an NPE. But do we want to silently ignore errors at this level?

**Comment by AllanWang** [comment]
@consp1racy 
1. Do you have links to the lines in question?
2. Are you still getting this error?

What you say may be true, but you'd be surprised how many core features seem to crash on specific devices. Hopefully a null check for a system service, even though it _should_ exist, may prevent some of the errors.

**Comment by consp1racy** [comment]
**TL;DR:** I've definitely gone off topic. The point is `ConnectivityManager` was not null so a null check did not resolve the error.

---

@AllanWang 

> Do you have links to the lines in question?

Sure, here are the links:

+ [`android.app.ContextImpl:289` in SDK 15](https://github.com/aosp-mirror/platform_frameworks_base/blob/df331873c8576e0ae34ae1ee3cc258beed373535/core/java/android/app/ContextImpl.java#L289)
+ [`android.app.SystemServiceRegistry:233` in SDK 25](https://github.com/aosp-mirror/platform_frameworks_base/blob/3de09838fb0996bb4b420630800ad34e828fd1b6/core/java/android/app/SystemServiceRegistry.java#L233) (sorry, I posted wrong line number earlier)

> Are you still getting this error?

I'm not getting the error, I'm just passing by. Just pointing out that your own crash report proves that **your `ConnectivityManager` was not null** therefore **the issue is still present**, just somewhere else.

> you'd be surprised how many core features seem to crash on specific devices [...] even though it should exist

I actually looked into this very topic about a month ago and divided system services into nullable and non-nullable according to AOSP source. [You can take a look at the list here.](https://github.com/consp1racy/android-commons/blob/master/commons-services/src/main/java/net/xpece/android/app/SystemServices.kt)

---

**EDIT**

Now I see that I didn't look far enough. Thanks for the impulse! `SystemServiceRegistry` may indeed return null if a native service was not found. I keep forgetting that. And services running on the device are determined by the [`SystemServer`](https://github.com/aosp-mirror/platform_frameworks_base/blob/master/services/java/com/android/server/SystemServer.java).

As you can find in `SystemServer` source absence of `ConnectivityManager` is considered a fatal error, unless...

The only way there is allowed no `ConnectivityManager` is that network would be disabled altogether in `/system/build.prop` by

    config.disable_network=true

---

But of course this is all pure Android. I agree that various OEMs strip/replace parts of the code.

On the other hand, look at it this way: You distribute your app through the Play Store. Play Store requires network to work. Network info is obtained from a `ConnectivityManager`. Therefore among your users there's no device without a `ConnectivityManager`. What are your thoughts on this?

**Comment by pwittchen** [comment]
I think, that problems mentioned by you guys are related to Android itself - not the library. Unfortunately, we cannot fix Android within this or another library. It may be good idea to post such issue to the Android issue tracker, so the source of the problem could be fixed.

**Comment by AllanWang** [comment]
@consp1racy even though I distribute through Play Store, I still can't guarantee the source of my app. I've had users build themselves and run on emulators. Others can also download an apk externally and copy it to a device. It doesn't look like any of this happened in this particular case, but it is certainly possible.

Given that this is not an issue within the library, I would recommend not implementing try catch. The goal of a library should not be to fix bugs outside of its scope, and we still have not had reports of NPEs after this fix.
